numbast/src/numbast/static/struct.py

```diff
@@ -0,0 +1,691 @@
+
+    def _render_data_model(self):
+        """Renders the data model of the struct."""
+
+        self.Imports.add("from numba.core.extending import register_model")
+
+        if self._data_model == PrimitiveModel:
+            self._data_model_rendered = self.primitive_data_model_template.format(
+                struct_type_name=self._struct_name,
+                struct_name=self._struct_name,
+            )
+        elif self._data_model == StructModel:
+            member_types_tuples = [
+                (f.name, to_numba_type(f.type_.unqualified_non_ref_type_name))
+                for f in self._decl.fields
+            ]
+
+            for _, nbty in member_types_tuples:
+                self.Imports.add(f"from numba.types import {nbty}")
+
+            member_types_tuples_strs = [
+                f"('{name}', {ty})" for name, ty in member_types_tuples
+            ]
+
+            member_types_str = ", ".join(member_types_tuples_strs)
+
+            self._data_model_rendered = self.struct_data_model_template.format(
+                struct_type_name=self._struct_name,
+                struct_name=self._struct_name,
+                member_types_tuples=member_types_str,
+            )
+
+    def _render_struct_attr(self):
+        """Renders the typings of the struct attributes."""
+
+        if self._data_model == StructModel:
+            self.Imports.add("from numba.cuda.cudadecl import register_attr")
+            self.Imports.add(
+                "from numba.core.typing.templates import AttributeTemplate"
+            )
+            self.Imports.add("from numba.core.extending import make_attribute_wrapper")
+
+            public_fields = [
+                f for f in self._decl.fields if f.access == access_kind.public_
+            ]
+
+            resolve_methods = []
+            attribute_wrappers = []
+            for field in public_fields:
+                resolve_methods.append(
+                    self.resolve_methods_template.format(
+                        attr_name=field.name,
+                        numba_type=to_numba_type(
+                            field.type_.unqualified_non_ref_type_name
+                        ),
+                    )
+                )
+                attribute_wrappers.append(
+                    self.make_attribute_wrappers_template.format(
+                        struct_name=self._struct_name,
+                        attr_name=field.name,
+                    )
+                )
+
+            resolve_methods_str = "\n".join(resolve_methods)
+            attribute_wrappers_str = "\n".join(attribute_wrappers)
+
+            self._struct_attr_typing_rendered = (
+                self.struct_attribute_typing_template.format(
+                    struct_name=self._struct_name,
+                    resolve_methods=resolve_methods_str,
+                    make_attribute_wrappers=attribute_wrappers_str,
+                )
+            )
+
+    def _render_struct_ctors(self):
```

Comment:
I'm not sure where best to place it, but we need to make sure we add imports for Numba types that correspond to the types passed in to constructors - e.g. if an object ctor takes a `double` we need to import `float64`, etc.

Reply:
Yes we do, I realized this in a separate PR. Will move some of the findings here.

Reply:
This is done in https://github.com/NVIDIA/numbast/pull/69/commits/618bf0724b9f4fb98922a8804f686cb3aa3ea469